File: GFG/Substrings of length k with k-1 distinct elements/Substrings of length k with k-1 distinct elements.cpp
class Solution {
  public:
    int countOfSubstrings(string S, int K) {
        int ans = 0, cnt_chars = 0;
        
        vector<int> v(26,0);
        for(int i=0;i<K;i++){
            int p = S[i] - 'a';
            if(v[p]==0) cnt_chars ++;
            v[p]++;
        }
        
        if(cnt_chars==K-1) ans++;
        
        for(int i=K;i<S.length();i++){
            int r = S[i] - 'a', l = S[i-K] - 'a';
            
            if(v[r]==0) cnt_chars++;
            v[r]++;
            
            if(v[l]==1) cnt_chars--;
            v[l]--;
            
            if(cnt_chars == K-1) ans++;
            // cout<<l<<" "<<r<<"\n";
        }
        
        return ans;
    }
};

File: GFG/Substrings of length k with k-1 distinct elements/Substrings of length k with k-1 distinct elements.java
class Solution {
    static int countOfSubstrings(String s, int k) {
        // code here
        int ans=0;
        int[]freq=new int[26];
        for(int i=0;i<k;i++){
            freq[s.charAt(i)-'a']++;
        }
        int cnt=0;
        for(int x:freq){
            if(x>0)cnt++;
        }
        int i=0;
        int j=k;
        while(j<=s.length()){
            if(cnt==k-1){
                ans++;
            }
            if(j==s.length())break;
            freq[s.charAt(i)-'a']--;
            if(freq[s.charAt(i)-'a']==0)cnt--;
            freq[s.charAt(j)-'a']++;
            if(freq[s.charAt(j)-'a']==1)cnt++;
            i++;
            j++;
        }
        
        return ans;
    }
};
